File: apps/api/app/schemas/dispatch.py
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, Field


DispatchType = Literal["incoming", "outgoing"]
DispatchStatus = Literal["draft", "registered", "processing", "completed", "archived"]


class DispatchBase(BaseModel):
    document_id: str
    dispatch_type: DispatchType
    document_number: str | None = Field(default=None, max_length=128)
    document_symbol: str | None = Field(default=None, max_length=128)
    issued_date: date | None = None
    issuing_agency: str | None = Field(default=None, max_length=255)
    recipient: str | None = None
    excerpt: str | None = None
    status: DispatchStatus = "draft"
    notes: str | None = None


class DispatchCreateRequest(DispatchBase):
    pass


class DispatchUpdateRequest(BaseModel):
    dispatch_type: DispatchType | None = None
    document_number: str | None = Field(default=None, max_length=128)
    document_symbol: str | None = Field(default=None, max_length=128)
    issued_date: date | None = None
    issuing_agency: str | None = Field(default=None, max_length=255)
    recipient: str | None = None
    excerpt: str | None = None
    status: DispatchStatus | None = None
    notes: str | None = None


class DispatchRead(BaseModel):
    id: str
    document_id: str
    document_title: str | None = None
    document_number: str | None = None
    document_status: str | None = None
    dispatch_type: str
    document_symbol: str | None = None
    issued_date: date | None = None
    issuing_agency: str | None = None
    recipient: str | None = None
    excerpt: str | None = None
    status: str
    notes: str | None = None
    created_at: datetime
    updated_at: datetime


class DispatchListResponse(BaseModel):
    items: list[DispatchRead]
    total: int
    limit: int
    offset: int

File: apps/web/types/dispatch.ts
export type DispatchType = 'incoming' | 'outgoing'
export type DispatchStatus = 'draft' | 'registered' | 'processing' | 'completed' | 'archived'

export interface DispatchItem {
  id: string
  document_id: string
  document_title?: string | null
  document_number?: string | null
  document_status?: string | null
  dispatch_type: DispatchType
  document_symbol?: string | null
  issued_date?: string | null
  issuing_agency?: string | null
  recipient?: string | null
  excerpt?: string | null
  status: DispatchStatus
  notes?: string | null
  created_at: string
  updated_at: string
}

export interface DispatchInput {
  document_id: string
  dispatch_type: DispatchType
  document_number?: string
  document_symbol?: string
  issued_date?: string
  issuing_agency?: string
  recipient?: string
  excerpt?: string
  status: DispatchStatus
  notes?: string
}

export interface DispatchListFilters {
  q?: string
  document_id?: string
  dispatch_type?: DispatchType | ''
  document_number?: string
  issuing_agency?: string
  status?: DispatchStatus | ''
  issued_date_from?: string
  issued_date_to?: string
  sort_by?: 'created_at' | 'updated_at' | 'document_number' | 'dispatch_type' | 'issuing_agency' | 'status' | 'issued_date'
  sort_dir?: 'asc' | 'desc'
  limit?: number
  offset?: number
}

export interface DispatchListResponse {
  items: DispatchItem[]
  total: number
  limit: number
  offset: number
}
